ansible_rulebook/websocket.py

```diff
@@ -12,126 +12,216 @@
+                    logger.info(
+                        "! connect failed; reconnecting in %.1f seconds",
+                        initial_delay,
+                        exc_info=True,
                     )
-                if data.get("type") == "ExtraVars":
-                    response.variables = yaml.safe_load(
-                        base64.b64decode(data.get("data"))
+                    await asyncio.sleep(initial_delay)
+                else:
+                    logger.info(
+                        "! connect failed again; retrying in %d seconds",
+                        int(backoff_delay),
+                        exc_info=True,
                     )
-                if data.get("type") == "ControllerInfo":
-                    response.controller_url = data.get("url")
-                    response.controller_token = data.get("token")
-                    response.controller_ssl_verify = data.get("ssl_verify")
-            return response
-        except CancelledError:
-            logger.info("closing websocket due to task cancelled")
-            return
-        except websockets.exceptions.ConnectionClosed:
-            logger.info("websocket %s closed", websocket_address)
-            return
-
-
-async def send_event_log_to_websocket(
-    event_log, websocket_address, websocket_ssl_verify
-):
-    logger.info("feedback websocket %s connecting", websocket_address)
-    event = None
-    async for websocket in websockets.connect(
-        websocket_address,
-        logger=logger,
-        ssl=_sslcontext(websocket_address, websocket_ssl_verify),
-    ):
-        logger.info("feedback websocket %s connected", websocket_address)
-        try:
-            if event:
-                logger.info("Resending last event...")
-                await websocket.send(json.dumps(event))
-                event = None
-
-            while True:
-                event = await event_log.get()
-                logger.debug(f"Event received, {event}")
-
-                if event == dict(type="Exit"):
-                    logger.info("Exiting feedback websocket task")
-                    return
-
-                await websocket.send(json.dumps(event))
-                event = None
-        except websockets.exceptions.ConnectionClosed:
-            logger.warning(
-                "feedback websocket %s connection closed, will retry...",
-                websocket_address,
+                    await asyncio.sleep(int(backoff_delay))
+                # Increase delay with truncated exponential backoff.
+                backoff_delay = backoff_delay * BACKOFF_FACTOR
+                backoff_delay = min(backoff_delay, BACKOFF_MAX)
+                continue
+        else:
+            # Connection succeeded - reset backoff delay
+            backoff_delay = BACKOFF_MIN
+            refresh = True
+
+    return result
+
+
+async def request_workload(activation_id: str) -> StartupArgs:
+    return await _connect_websocket(
```

Comment:
@bzwei when the passed in token is bad will the program end or would it try the refresh token?

Reply:
In this case it will try refreshing once, and then re-throw the exception if it doesn't help.